### PR TITLE
Added ability to define service type into pipeline specs

### PR DESCRIPTION
**Pull request by DanielMorales9**
As per title, I added the ability to declare the service-type in the service section of the pipeline spec: pachyderm/pachyderm#3650
````

"service": {
    "internal_port": int,
    "external_port": int,
    "type": "LoadBalancer|ClusterIP|NodePort"
  },
`````
The available service types are LoadBalancer, ClusterIP and NodePort (which is the default choice if  service-type is not declared). 
I am not supporting ExternalName service type as it would require me to define an extra field, i.e. externalName.

### CODE REVIEW

**Comment by DanielMorales9** [comment]
What is the motivation of running ``protoc`` locally and then pushing it to origin? Shouldn't be the compilation part of the CI phase?

File: src/server/pps/server/worker_rc.go [hunk 1]
```diff
@@ -490,15 +506,8 @@ func (a *apiServer) createWorkerRc(options *workerOptions) error {
...
-				Type:     v1.ServiceTypeNodePort,
-				Ports: []v1.ServicePort{
-					{
-						Port:       options.service.ExternalPort,
-						TargetPort: intstr.FromInt(int(options.service.InternalPort)),
-						Name:       "user-port",
-						NodePort:   options.service.ExternalPort,
-					},
-				},
+				Type:     serviceType,
```
Comment (jdoliner) [inline src/server/pps/server/worker_rc.go]: Small tweak: I think you can just do `Type: options.service.Type` and save the conditional above, with the exception of the part where you set the `NodePort`. `v1.ServiceType` is actually just a string under the hood so you can assign to it directly.

File: src/server/pps/server/api_server.go [hunk 2]
```diff
@@ -60,6 +60,8 @@ const (
 	// DefaultDatumTries is the default number of times a datum will be tried
 	// before we give up and consider the job failed.
 	DefaultDatumTries = 3
+	// DefaultServiceType is the default service type for a service pipeline
+	DefaultServiceType = "NodePort"
```
Comment (jdoliner) [inline src/server/pps/server/api_server.go]: Could we use, `v1.ServiceTypeNodePort` here in place of hard coding the value.

File: src/server/pps/server/api_server.go [hunk 3]
```diff
@@ -1540,6 +1542,13 @@ func (a *apiServer) validatePipeline(pachClient *client.APIClient, pipelineInfo
 	if pipelineInfo.PodPatch != "" && !json.Valid([]byte(pipelineInfo.PodPatch)) {
 		return fmt.Errorf("malformed PodPatch")
 	}
+	if pipelineInfo.Service != nil {
+		validServiceTypes := map[string]bool{"ClusterIP": true, "LoadBalancer": true, "NodePort": true}
```
Comment (jdoliner) [inline src/server/pps/server/api_server.go]: Similar here, using `v1.ServiceTypeClusterIP`, etc.

**Comment by DanielMorales9** [comment]
Thanks, @jdoliner, when do you think this can be released? Should I just follow the custom release docs. Many thanks for support

**Comment by gabrielgrant** [comment]
Hey @DanielMorales9 thanks for the PR, and sorry for the slow reply. This may already be out in the latest 1.9 release candidate, but if not, it should be included in the next RC (which I expect will be shipped in the next day or two)

**Pull request merged**